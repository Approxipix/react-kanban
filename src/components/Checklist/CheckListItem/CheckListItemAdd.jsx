import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addCheckListItem } from '../../../redux/checkListReducer/actions/actions';
import { Actions, Input, SubmitButton, CancelButton } from '../../BaseComponent';
import ClickOutside from "../../ClickOutside/ClickOutside";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Wrapper = styled.div`
  padding-left: 2rem;
`;

const AddButton = styled.button`
  padding: .5rem 1rem;
  font-size: .875rem;
  color: #172b4d;
  background-color: rgba(9, 30, 66, .08);
  border-radius: 3px;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;
AddButton.displayName = 'AddButton';

class CheckListItemAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      title: '',
    })
  };

  handleChange = (event) => {
    this.setState({
      title: event.target.value,
    });
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.toggleOpened();
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.addCheckListItem({
      checkListId: this.props.checkListId,
      checkListTitle: title,
    });
    this.toggleOpened()
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Wrapper>
          <form>
            <Input
              type="text"
              size=".875rem"
              padding=".5rem"
              placeholder="Enter a card title..."
              value={title}
              onKeyDown={this.handleKeyDown}
              onChange={(e) => this.handleChange(e)}
              spellCheck={false}
              autoFocus
            />
            <Actions>
              <SubmitButton onClick={(e) => this.handleSubmit(e)} disabled={!title}>
                Save
              </SubmitButton>
              <CancelButton onClick={() => this.toggleOpened()}>
                <FontAwesomeIcon icon="times" />
              </CancelButton>
            </Actions>
          </form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <Wrapper>
        <AddButton onClick={() => this.toggleOpened()}>
          Add a item
        </AddButton>
      </Wrapper>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addCheckListItem: addCheckListItem,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CheckListItemAdd);
