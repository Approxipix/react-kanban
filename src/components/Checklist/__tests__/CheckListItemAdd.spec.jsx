import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CheckListItemAdd from '../CheckListItem/CheckListItemAdd';

const mockStore = configureStore([]);

describe('<CheckListItemAdd>', () => {
  let store;
  let component;
  let mockAddCheckListItem = jest.fn();

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = mockAddCheckListItem;
    const props = {
      checkListId: 'CheckListID',
    };
    const state = {
      isOpened: false,
      title: '',
    };

    const wrapper = shallow(<CheckListItemAdd store={store} {...props} />);
    component = wrapper.find('CheckListItemAdd').dive();
    component.setState({...state})
  });

  it('should render AddButton if isOpened is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render form if isOpened is true', () => {
    component.instance().toggleOpened();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleOpened sets isOpened correctly', () => {
    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(true);

    component.instance().toggleOpened();
    expect(component.instance().state.isOpened).toEqual(false);
  });

  it("should call toggleOpened after click on AddButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.find('AddButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should call toggleOpened after click on CancelButton", () => {
    const spy = jest.spyOn(component.instance(), "toggleOpened");

    component.setState({ isOpened: true });
    component.find('CancelButton').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });

  describe('add checklist item form', () => {
    beforeEach(() => {
      component.setState({ isOpened: true });
    });

    it("should call toggleOpened if key code equal 27", () => {
      const spy = jest.spyOn(component.instance(), "toggleOpened");

      component.instance().handleKeyDown({ keyCode: 27 });

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call handleChange on boardTitle change with the correct params', () => {
      const spy = jest.spyOn(component.instance(), "handleChange");
      const mockEvent = {
        target: {
          name: "title",
          value: "Test title"
        }
      };

      component.find("Input").simulate("change", mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(mockEvent);
    });

    it('should call setState on title', () => {
      const mockEvent = {
        target: {
          name: "title",
          value: "Test Title"
        }
      };
      const expectedState = {
        title: "Test Title",
        isOpened: true
      };

      component.instance().handleChange(mockEvent);

      expect(component.state()).toEqual(expectedState)
    });

    it("should call handleSubmit with the correct params", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.setState({ title: 'Test Title' });
      component.instance().handleSubmit(mockEvent);

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should return handleSubmit if title is empty", () => {
      const spy = jest.spyOn(component.instance(), "handleSubmit");
      const mockEvent = {
        preventDefault: jest.fn()
      };

      component.instance().handleSubmit(mockEvent);

      expect(spy).toReturn();
    });

    it("should dispatch an action after handleSubmit", () => {
      const expectedAction = {
        type: 'ADD_CHECKLIST_ITEM',
        payload: {
          checkListId: 'CheckListID',
          checkListTitle: 'Test Title',
        }
      };

      component.setState({ title: 'Test Title' });
      component.find('SubmitButton').simulate('click');

      expect(mockAddCheckListItem).toHaveBeenCalledTimes(1);
      expect(mockAddCheckListItem).toHaveBeenCalledWith(expectedAction);
    });
  })
});
