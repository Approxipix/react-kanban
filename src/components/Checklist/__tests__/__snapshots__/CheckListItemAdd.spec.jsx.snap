// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckListItemAdd> should render AddButton if isOpened is false 1`] = `
<styled.div>
  <AddButton
    onClick={[Function]}
  >
    Add a item
  </AddButton>
</styled.div>
`;

exports[`<CheckListItemAdd> should render form if isOpened is true 1`] = `
<ClickOutside
  toggleOpened={[Function]}
>
  <styled.div>
    <form
      onSubmit={[Function]}
    >
      <Input
        autoFocus={true}
        onChange={[Function]}
        onKeyDown={[Function]}
        padding=".5rem"
        placeholder="Enter a card title..."
        size=".875rem"
        spellCheck={false}
        type="text"
        value=""
      />
      <styled.div>
        <SubmitButton
          disabled={true}
          type="submit"
        >
          Save
        </SubmitButton>
        <CancelButton
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="times"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </CancelButton>
      </styled.div>
    </form>
  </styled.div>
</ClickOutside>
`;
